### PR TITLE
fix(zoe,pegasus): some zoe and pegasus invitations use proposal patterns

### PR DESCRIPTION
**Pull request by erights**
First commit was extracted from #7041 . Likely some of the review comments there are relevant here.

Replaces some uses of the deprecated `assertProposalShape` with `makeInvitation`s direct support for proposalShape patterns. This provides better protection for contracts against malformed offers. When an offer's proposal violates the contract's required proposalShape pattern, it is rejected in Zoe and the contract never even sees the malformed offer.

This PR is somewhat empirical, revising the provided proposal patterns so that they were tighter than the `assertProposalPattern` call it replaced, while still passing all tests. For each modified contract, it would be good for someone who knows the meaning of the contract to do a sanity check on the proposalShape pattern.

One particular way these patterns could be tighter, to protect the contract from seeing more malformed offers, is to use the amountShape specific to the brand the contract is willing to accept in that position. I skipped that additional work, leaving a TODO there instead. Reviewers, if you know what brand-specific amountShape I should use in any such case, please let me know.

# Upgrade considerations

This PR was extracted from #7041 to omit packages/governance, so that we can worry separately about its upgrade sensitivity.

### CODE REVIEW

File: packages/zoe/src/contractSupport/zoeHelpers.js [hunk 1]
```diff
@@ -199,6 +200,12 @@ export const depositToSeat = async (zcf, recipientSeat, amounts, payments) => {
   const invitation = zcf.makeInvitation(
     reallocateAfterDeposit,
     'temporary seat for deposit',
+    undefined,
+    M.splitRecord({
+      give: AmountKeywordRecordShape,
+      want: {},
+      exit: { onDemand: null },
+    }),
```
Comment (dckc) [inline packages/zoe/src/contractSupport/zoeHelpers.js:208]: This `zoeHelpers.js` file seems to be part of a number of production bundles:

 - `scaledPriceAuthority`
 - `fluxAggregatorKit`
 - `vaultFactory`
 - `reserve`
 - `psm`

#8219 seems to represent plans to test upgrading all those. We could optimistically land this, presuming it'll all work out. Or we could block this until those tests are in place. @turadg preferences? This seems fairly low risk.

<details><summary>bundle audit details</summary>

```
agoric-sdk$ find . -name 'bundle-*-meta.js' | xargs grep contractSupport/zoeHelpers.js
./packages/smart-wallet/bundles/bundle-walletFactory-meta.js:      "relativePath": "../../../zoe/src/contractSupport/zoeHelpers.js",
./packages/inter-protocol/bundles/bundle-scaledPriceAuthority-meta.js:      "relativePath": "../../contractSupport/zoeHelpers.js",
./packages/inter-protocol/bundles/bundle-fluxAggregatorKit-meta.js:      "relativePath": "../../../../zoe/src/contractSupport/zoeHelpers.js",
./packages/inter-protocol/bundles/bundle-vaultFactory-meta.js:      "relativePath": "../../../../zoe/src/contractSupport/zoeHelpers.js",
./packages/inter-protocol/bundles/bundle-reserve-meta.js:      "relativePath": "../../../../zoe/src/contractSupport/zoeHelpers.js",
./packages/inter-protocol/bundles/bundle-psm-meta.js:      "relativePath": "../../../../zoe/src/contractSupport/zoeHelpers.js",
./packages/inter-protocol/bundles/bundle-auctioneer-meta.js:      "relativePath": "../../../../zoe/src/contractSupport/zoeHelpers.js",
./packages/zoe/bundles/bundle-priceAggregator-meta.js:      "relativePath": "../../contractSupport/zoeHelpers.js",
./packages/zoe/bundles/bundle-auction-meta.js:      "relativePath": "../../../contractSupport/zoeHelpers.js",
./packages/governance/bundles/bundle-committee-meta.js:      "relativePath": "../../../zoe/src/contractSupport/zoeHelpers.js",
./packages/pegasus/bundles/bundle-pegasus-meta.js:      "relativePath": "../../../zoe/src/contractSupport/zoeHelpers.js",
```

</details>

**Review by dckc — COMMENTED** [review]
looking for input on whether this should wait for #8219

File: packages/zoe/src/contracts/auction/index.js [hunk 2]
```diff
@@ -116,15 +116,33 @@ const start = zcf => {
     });
   };
 
+  const BidProposalShape = M.splitRecord({
```
Comment (Chris-Hibbert) [inline packages/zoe/src/contracts/auction/index.js:119]: I like `makeNatAmountShape()` from `packages/inter-protocol/src/contractSupport.js`.

File: packages/pegasus/src/pegasus.js [hunk 3]
```diff
@@ -20,11 +19,11 @@ import { makeCourierMaker, getCourierPK } from './courier.js';
 const DEFAULT_DENOM_TRANSFORMER = IBCSourceTraceDenomTransformer;
 const DEFAULT_TRANSFER_PROTOCOL = ICS20TransferProtocol;
 
-const TRANSFER_PROPOSAL_SHAPE = {
+const TransferProposalShape = M.splitRecord({
   give: {
-    Transfer: null,
+    Transfer: AmountShape, // TODO get amount shape from brand
```
Comment (Chris-Hibbert) [inline packages/pegasus/src/pegasus.js:24]: I like `makeNatAmountShape(brand)` for now.

**Review by dckc — APPROVED** [review]
same justification for landing #7041 before an upgrade test seems to apply here.

File: packages/pegasus/src/pegasus.js (hunk 3 above)
Comment (erights) [reply packages/pegasus/src/pegasus.js:24]: Hi @Chris-Hibbert (cc @michaelfig since it's pegasus)
TIL `makeNatAmountShape`. Nice!

But from a brief skim of the code, I did not figure out what expression to write in this location for the brand. ELI5, could you show me plausible code? Thanks.

File: packages/zoe/src/contracts/auction/index.js (hunk 2 above)
Comment (erights) [reply packages/zoe/src/contracts/auction/index.js:119]: From a brief skim of the code, I did not figure out what expression to write in this location for the brand. ELI5, could you show me plausible code? Thanks.

File: packages/pegasus/src/pegasus.js (hunk 3 above)
Comment (dckc) [reply packages/pegasus/src/pegasus.js:24]: `denomState.localBrand` seems plausible.

Given the uncertainty, it seems worthwhile to add tests:
 - offers that use the right brand work
 - offers that use the wrong brand don't work

File: packages/zoe/src/contracts/auction/index.js (hunk 2 above)
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/auction/index.js:119]: The brands aren't known until the sell invitation is exercised, so it is correct for `SellProposalShape` to stay generic. After that, we can create the bid shape:
```
    const bidBrand = sellSeat.getProposal().want.Ask.brand;
    const AmountShape = makeNatAmountShape(bidBrand);

    const BidProposalShape = M.splitRecord({
      give: { Bid: AmountShape },
      want: { Asset: AmountShape },
    });
```
`BidProposalShape` is then used in `makeBidInvitation`. We can either use `let` to set a global, or use currying to build the `makeBidInvitation` that is returned to the seller. I think I prefer the latter, so the first line of `makeBidInvitation` changes to 
```
  const buildBidInvitationMaker = bidProposalShape => {
    return () => {
```
and we have to insert this line before the return in `sell`.
```
    const makeBidInvitation = buildBidInvitationMaker(BidProposalShape);
```

**Review by turadg — CHANGES_REQUESTED** [review]
Requesting that:
- Upgrade Considerations specifies how this could break Mainnet and when in what release it belongs
- transfer uses `makeNatAmountShape`
- Regression tests are added for what this is meant to fix
